File: clients/ts/src/baml_lib/deserializer/base_deserializer.ts
import { JSONSchema7 } from "json-schema";
import { Diagnostics } from "./diagnostics";
import { RawWrapper } from "./raw_wrapper/raw_wrapper";

class Result<T> {
    has_value: boolean;
    value?: T;

    private constructor(has_value: boolean, value?: T) {
        this.has_value = has_value;
        this.value = value;
    }

    static from_value<T>(value: T): Result<T> {
        return new Result(true, value);
    }

    static failed<T>(): Result<T> {
        return new Result<T>(false, undefined);
    }

    get as_value(): T {
        if (!this.has_value) {
            throw new Error("Result does not have a value");
        }
        return this.value as T;
    }
}

type CheckLutFn<T> = (def: JSONSchema7) => BaseDeserializer<T>;

abstract class BaseDeserializer<T> {
    private _rank: number;

    constructor(rank: number) {
        this._rank = rank;
    }

    get rank() {
        return this._rank;
    }
    abstract coerce(raw: RawWrapper, diagnostics: Diagnostics, fromLut: CheckLutFn<any>): Result<T>;
}

export { BaseDeserializer, Result, CheckLutFn };

File: clients/python/baml_lib/_impl/deserializer/base_deserializer.py
import abc
import typing


from .diagnostics import Diagnostics, DeserializerError, DeserializerWarning
from .raw_wrapper import RawWrapper
from .type_definition import ITypeDefinition

T = typing.TypeVar("T")


class Result(typing.Generic[T]):
    has_value: bool
    value: typing.Optional[T]

    def __init__(self, *, has_value: bool, value: typing.Optional[T]) -> None:
        self.has_value = has_value
        self.value = value

    @staticmethod
    def from_value(value: T) -> "Result[T]":
        return Result(has_value=True, value=value)

    @staticmethod
    def failed() -> "Result[T]":
        return Result(has_value=False, value=None)

    @property
    def as_value(self) -> T:
        assert self.has_value, "Result has no value"
        # Technically self.value could be None, since T could itself be Optional.
        return typing.cast(T, self.value)


CoerceFn = typing.Callable[
    [
        RawWrapper,
        Diagnostics,
        typing.Callable[[str], typing.Optional["BaseDeserializer[T]"]],
    ],
    Result[T],
]


class BaseDeserializer(typing.Generic[T], metaclass=abc.ABCMeta):
    def __init__(self, rank: int) -> None:
        self.__rank = rank

    @property
    def rank(self) -> int:
        # Rank means lower is less specific, higher is more specific.
        # When searching for a deserializer, we want to pick the most specific.
        return self.__rank

    @abc.abstractmethod
    def coerce(
        self,
        raw: RawWrapper,
        diagnostics: Diagnostics,
        from_lut: typing.Callable[
            [ITypeDefinition], "typing.Optional[BaseDeserializer[typing.Any]]"
        ],
    ) -> Result[T]:
        raise NotImplementedError()


CheckLutFn = typing.Callable[[ITypeDefinition], typing.Optional[BaseDeserializer[T]]]


__all__ = [
    "BaseDeserializer",
    "Result",
    "Diagnostics",
    "DeserializerError",
    "DeserializerWarning",
    "RawWrapper",
    "CoerceFn",
    "ITypeDefinition",
]
